src/client/mapblock_mesh.cpp

```diff
@@ -585,6 +587,47 @@ void PartialMeshBuffer::draw(video::IVideoDriver *driver) const
 	MapBlockMesh
 */
 
+static void applyColorAndMerge(std::vector<PreMeshBuffer> &prebuffers)
+{
+	// TODO: we should change the meshgen so it already applies the tile color
+	// so that we don't need to this extra step.
+	// However currently the CAO code relies on the ability to erase the vertex
+	// colors (light data) before applying the tile colors.
+
+	for (auto &p : prebuffers) {
+		p.applyTileColor();
+		// erase color information for later comparisons
+		p.layer.has_color = false;
+		p.layer.color = 0;
+	}
+
+	std::unordered_map<TileLayer, size_t> seen;
+	for (size_t i = 0; i < prebuffers.size(); i++) {
+		PreMeshBuffer &p = prebuffers[i];
+		auto it = seen.find(p.layer);
+		if (it == seen.end()) { // first time
+			seen[p.layer] = i;
+			continue;
+		}
+		// merge
+		auto &dst = prebuffers[it->second];
+		assert(p.layer == dst.layer);
+		if (dst.append(p)) {
+			p = PreMeshBuffer();
+		} else {
+			// other buffer full, this one becomes the new target
+			it->second = i;
+		}
+	}
+
+	{
+		auto &c = prebuffers;
```

Comment:
1. This does IMO not improve readability and would prefer the more verbose notation - which also happens to make the separate code block unneeded.
2. Wouldn't it be cheaper to leave those empty entries, and skip them in the `MapBlockMesh::MapBlockMesh` loop?

Reply:
> Wouldn't it be cheaper to leave those empty entries, and skip them in the MapBlockMesh::MapBlockMesh loop?

I tried that, but it breaks things. I think there's some reliance on the buffer indices in the final mesh. :shrug:

Reply:
Hm that's strange but acceptable. Then let's keep it like this. Thank you for the comment inside the loop.